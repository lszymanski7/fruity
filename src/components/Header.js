import React, { useState } from 'react'
import ToggleSwitch from './ToggleSwitch'
import { logos } from '../data/constants'

const Header = () => {
    // Application name
    const name = 'FRUITY'

    // Toggle switch states for dark and light mode
    const [isChecked, setIsChecked] = useState(false)

    return (
        <header className="flexbox-column header">
            <div className="flexbox-row header__widgets">
                <a href="https://github.com/lszymanski7/fruity-app">
                    <img
                        alt="GitHub Logo"
                        className="header__github-logo"
                        src={logos.github}
                    />
                </a>
                <ToggleSwitch
                    handleOnChange={() => setIsChecked(!isChecked)}
                    icons={['moon-icon', 'sun-icon']}
                    id="themeSwitcher"
                    isChecked={isChecked}
                />
            </div>
            <div className="flexbox-row header__brand">
                <img
                    alt="Fruity Logo"
                    className="header__app-logo"
                    src={logos.fruity.size_256x256}
                />
                <h1 className="header__app-name">{name}</h1>
            </div>
        </header>
    )
}

export { Header as default }
