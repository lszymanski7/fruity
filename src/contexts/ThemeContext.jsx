/* istanbul ignore file */

import React, { createContext, useEffect } from 'react'
import PropTypes from 'prop-types'
import useLocalStorage from '../hooks/useLocalStorage'

// Function ➞ creates theme context
export const ThemeContext = createContext()

// Function ➞ creates theme provider
export const ThemeProvider = ({ children }) => {
  const [theme, setTheme] = useLocalStorage('theme', 'light')

  const toggleTheme = () => {
    const nextState = theme === 'dark' ? 'light' : 'dark'
    setTheme(nextState)
  }

  useEffect(() => {
    switch (theme) {
      case 'dark':
        document.body.className = 'theme-dark'
        break
      case 'light':
        document.body.className = 'theme-light'
        break
      default:
        throw new Error(`Unhandled theme: ${theme}`)
    }
  }, [theme])

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  )
}

ThemeProvider.propTypes = {
  children: PropTypes.object.isRequired
}
