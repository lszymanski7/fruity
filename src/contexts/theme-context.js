/* istanbul ignore file */

import React, { createContext, useEffect } from 'react'
import PropTypes from 'prop-types'
import useLocalStorage from '../hooks/useLocalStorage'

// Function ➞ creates theme context
const ThemeContext = createContext()

// Function ➞ creates theme provider
const ThemeProvider = ({ children }) => {
    const [theme, setTheme] = useLocalStorage('theme', 'light')

    const toggleTheme = () => {
        const nextState = theme === 'dark' ? 'light' : 'dark'
        setTheme(nextState)
    }

    useEffect(() => {
        switch (theme) {
            case 'dark':
                document.body.className = 'background-dark'
                break
            case 'light':
                document.body.className = 'background-light'
                break
            default:
                throw new Error(`Unhandled theme: ${theme}`)
        }
    }, [theme])

    return (
        <ThemeContext.Provider value={{ theme, toggleTheme }}>
            {children}
        </ThemeContext.Provider>
    )
}

ThemeProvider.propTypes = {
    children: PropTypes.object.isRequired
}

export { ThemeContext, ThemeProvider }
