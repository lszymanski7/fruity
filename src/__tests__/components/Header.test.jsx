import React from 'react'
import { render, screen } from '@testing-library/react'
import Header from '../../components/Header'
import { ThemeProvider } from '../../contexts/ThemeContext'

describe('Header', () => {
    it('should be rendered correctly.', () => {
        const { container } = render(<Header />, { wrapper: ThemeProvider })
        expect(container).toMatchSnapshot()
    })

    it('should have the GitHub logo.', () => {
        render(<Header />, { wrapper: ThemeProvider })
        const img = screen.getAllByRole('img')[0]
        expect(img).toHaveAttribute('alt', 'GitHub Logo')
        expect(img).toHaveAttribute('src', 'github.svg')
    })

    it('should have link with the correct href attribute.', () => {
        render(<Header />, { wrapper: ThemeProvider })
        const link = screen.getByRole('link')
        expect(link).toHaveAttribute('href', 'https://github.com/lszymanski7/fruity-app')
    })

    it('should have a toggle switch (dark/light mode).', () => {
        render(<Header />, { wrapper: ThemeProvider })
        const themeSwitch = screen.getByTestId('theme-switch')
        expect(themeSwitch).toBeInTheDocument()
    })

    it('should have the correct application logo.', () => {
        render(<Header />, { wrapper: ThemeProvider })
        const img = screen.getAllByRole('img')[1]
        expect(img).toHaveAttribute('alt', 'Application Logo')
        expect(img).toHaveAttribute('src', 'fruity-256x256.webp')
    })

    it('should have the correct application name.', () => {
        render(<Header />, { wrapper: ThemeProvider })
        const name = 'FRUITY'
        const h1 = screen.getByRole('heading', { level: 1 })
        expect(h1).toHaveTextContent(name)
    })
})
