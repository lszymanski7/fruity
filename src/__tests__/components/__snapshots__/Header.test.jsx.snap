// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should be rendered correctly. 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="header__widgets-container"
    >
      <a
        href="https://github.com/lszymanski7/fruity-app"
      >
        <img
          alt="GitHub Logo"
          class="header__github-logo header__github-logo--light"
          src="github.svg"
        />
      </a>
      <div
        class="toggle-switch"
        data-testid="theme-switch"
      >
        <input
          class="toggle-switch__input"
          id="theme-switch"
          type="checkbox"
        />
        <label
          class="toggle-switch__label"
          for="theme-switch"
        >
          <span
            class="toggle-switch__icon moon-icon"
            data-testid="moon"
          />
          <span
            class="toggle-switch__icon sun-icon"
            data-testid="sun"
          />
          <span
            class="toggle-switch__slider"
          />
        </label>
      </div>
    </div>
    <div
      class="header__brand-container"
    >
      <img
        alt="Application Logo"
        class="header__app-logo"
        src="fruity-256x256.webp"
      />
      <h1
        class="header__app-name"
      >
        FRUITY
      </h1>
    </div>
  </header>
</div>
`;
